### PR TITLE
Added support for Razer Blackwidow V3 Mini Hyperspeed

### LINKED ISSUE #1544: Support for BlackWidow V3 Mini Hyperspeed
### Device Info

* Name: BlackWidow V3 mini hyperspeed
* Product Number: RZ03-03890300
* Model Number: RZ03-0389

### Device Information
```
Please post the output of
    lsusb -d 1532: -v
```
in macos i have no lusb, but device id is 0x0271 via wireless dongle, 0x258 connected with the usb wire
```
Please post the output of
    ls -lah /dev/input/by-id/
```

### Packet Captures

<Please see here as we probably need packet captures>
<https://github.com/openrazer/openrazer/wiki/Reverse-Engineering-USB-Protocol>

### Synapse Screenshots
<Screenshots of all available options as seen on Razer Synapse>
this is the one : 

https://www.razer.com/gaming/keyboards/razer-blackwidow-v3-mini-hyperspeed

### PR DESCRIPTION
**Pull request by IvanVojtko**
Hey!
This PR adds support for Razer Blackwidow V3 Mini Hyperspeed (wired and wireless connection). I've also added support to report battery status. 
Fixes https://github.com/openrazer/openrazer/issues/1544

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -77,6 +77,8 @@ The devices below are fully feature supported by OpenRazer, which means all avai
 | Razer Blade 15 Advanced (Early 2021)          |  1532:026D  |
 | Razer Blade 15 Advanced (Mid 2021)            |  1532:0276  |
 | Razer BlackWidow V3 Tenkeyless                |  1532:0A24  |
+| Razer BlackWidow V3 Mini Hyperspeed           |  1532:0258  |
+| Razer BlackWidow V3 Mini Hyperspeed Wireless  |  1532:0271  |
```
Comment (z3ntu) [inline README.md]: Please keep this list sorted by USB PID

File: daemon/openrazer_daemon/hardware/keyboards.py [hunk 2]
```diff
@@ -759,6 +760,78 @@ class RazerBlackWidowV3TK(_RippleKeyboard):
...
+    def _close(self):
+        """
+        Close the key manager
+        """
+        super()._close()
+
+        self._battery_manager.close()
+
+
+class RazerBlackWidowV3MINIWireless(_RippleKeyboard):
```
Comment (z3ntu) [inline daemon/openrazer_daemon/hardware/keyboards.py]: You can make this class inherit from `RazerBlackWidowV3MINI` (or the other way around) and remove the common properties so just keep things like USB_PID etc

File: driver/razerkbd_driver.c [hunk 3]
```diff
@@ -285,13 +327,42 @@ static void razer_set_device_mode(struct usb_device *usb_dev, unsigned char mode
         report.transaction_id.id = 0x3F;
         break;
     case USB_DEVICE_ID_RAZER_BLACKWIDOW_ELITE:
+    case USB_DEVICE_ID_RAZER_BLACKWIDOW_V3_MINI:
         report.transaction_id.id = 0x1F;
         break;
+    case USB_DEVICE_ID_RAZER_BLACKWIDOW_V3_MINI_WIRELESS:
+        report.transaction_id.id = 0x9F;
+        break;
```
Comment (z3ntu) [inline driver/razerkbd_driver.c:342]: Interesting that they use different transaction ids... Well if it works it's okay and 0x9f also makes sense :D

File: driver/razerkbd_driver.c [hunk 4]
```diff
@@ -285,13 +327,42 @@ static void razer_set_device_mode(struct usb_device *usb_dev, unsigned char mode
...
     }
-
```
Comment (z3ntu) [inline driver/razerkbd_driver.c:292]: unnecessary change

File: driver/razerkbd_driver.c [hunk 5]
```diff
@@ -2563,6 +2739,23 @@ static int razer_kbd_probe(struct hid_device *hdev, const struct hid_device_id *
...
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_matrix_effect_none);            // No effect
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_matrix_effect_reactive);        // Reactive effect
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_matrix_effect_breath);          // Breathing effect
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_matrix_effect_static);          // Static effect
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_matrix_effect_custom);          // Custom effect
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_matrix_custom_frame);           // Set LED matrix
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_game_led_state);                // Enable game mode & LED
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_macro_led_state);               // Enable macro LED
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_macro_led_effect);              // Change macro LED effect (static, flashing)
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_charge_level);              // Change macro LED effect (static, flashing)
```
Comment (z3ntu) [inline driver/razerkbd_driver.c]: nitpick: align comment on this line with other comments

File: driver/razerkbd_driver.c [hunk 6]
```diff
@@ -2909,6 +3102,23 @@ static void razer_kbd_disconnect(struct hid_device *hdev)
...
+            device_remove_file(&hdev->dev, &dev_attr_matrix_effect_none);            // No effect
+            device_remove_file(&hdev->dev, &dev_attr_matrix_effect_reactive);        // Reactive effect
+            device_remove_file(&hdev->dev, &dev_attr_matrix_effect_breath);          // Breathing effect
+            device_remove_file(&hdev->dev, &dev_attr_matrix_effect_static);          // Static effect
+            device_remove_file(&hdev->dev, &dev_attr_matrix_effect_custom);          // Custom effect
+            device_remove_file(&hdev->dev, &dev_attr_matrix_custom_frame);           // Set LED matrix
+            device_remove_file(&hdev->dev, &dev_attr_game_led_state);                // Enable game mode & LED
+            device_remove_file(&hdev->dev, &dev_attr_macro_led_state);               // Enable macro LED
+            device_remove_file(&hdev->dev, &dev_attr_macro_led_effect);              // Change macro LED effect (static, flashing)
+            device_remove_file(&hdev->dev, &dev_attr_charge_level);
```
Comment (z3ntu) [inline driver/razerkbd_driver.c]: nitpick: add comment like with CREATE_DEVICE_FILE

File: driver/razerkbd_driver.h [hunk 7]
```diff
@@ -80,6 +80,8 @@
 #define USB_DEVICE_ID_RAZER_BLADE_15_ADV_EARLY_2021 0x026D
 #define USB_DEVICE_ID_RAZER_BLADE_15_ADV_MID_2021 0x0276
 #define USB_DEVICE_ID_RAZER_BLACKWIDOW_V3_TK 0x0A24
+#define USB_DEVICE_ID_RAZER_BLACKWIDOW_V3_MINI 0x0258
+#define USB_DEVICE_ID_RAZER_BLACKWIDOW_V3_MINI_WIRELESS 0x0271
```
Comment (z3ntu) [inline driver/razerkbd_driver.h]: please keep the entries sorted by value

File: driver/razerkbd_driver.h [hunk 8]
```diff
@@ -120,7 +122,7 @@
 
 #define RAZER_BLACKWIDOW_CHROMA_WAIT_MS 1
 #define RAZER_BLACKWIDOW_CHROMA_WAIT_MIN_US 600
-#define RAZER_BLACKWIDOW_CHROMA_WAIT_MAX_US 800
+#define RAZER_BLACKWIDOW_CHROMA_WAIT_MAX_US 5000
```
Comment (z3ntu) [inline driver/razerkbd_driver.h]: You should introduce new values for your device, see razermouse_driver.c how it's done there

Also these values don't make much sense, it's a usleep that's actually sleeping in a more or less random value between these two numbers, so every time it's sleeping some time between 600 and 5000 nano(?)seconds; so min & max are best kept 100 apart

File: install_files/udev/99-razer.rules [hunk 9]
```diff
@@ -10,7 +10,7 @@ ATTRS{idProduct}=="0013|0016|0020|0024|0025|002e|002f|0032|0034|0036|0037|0038|0
     ENV{ID_RAZER_CHROMA}="1", ENV{RAZER_DRIVER}="razermouse"
 
 # Keyboards
-ATTRS{idProduct}=="010d|010e|010f|0111|0113|011a|011b|011c|0201|0202|0203|0204|0205|0207|0208|0209|020f|0210|0211|0214|0216|0217|021a|021e|021f|0220|0221|0224|0225|0226|0227|0228|022a|022b|022c|022d|022f|0232|0233|0234|0235|0237|0239|023a|023b|023f|0240|0241|0243|0245|0246|024a|024c|024d|0252|0253|0255|0257|0259|025d|025e|0266|026a|026d|0276|0a24", \
+ATTRS{idProduct}=="010d|010e|010f|0111|0113|011a|011b|011c|0201|0202|0203|0204|0205|0207|0208|0209|020f|0210|0211|0214|0216|0217|021a|021e|021f|0220|0221|0224|0225|0226|0227|0228|022a|022b|022c|022d|022f|0232|0233|0234|0235|0237|0239|023a|023b|023f|0240|0241|0243|0245|0246|024a|024c|024d|0252|0253|0255|0257|0259|025d|025e|0266|026a|026d|0276|0a24|0258|0271", \
```
Comment (z3ntu) [inline install_files/udev/99-razer.rules]: keep this list sorted as well

File: daemon/resources/razer.conf [hunk 10]
```diff
@@ -18,3 +18,9 @@ mouse_battery_notifier_freq = 600
 
 # Apply effects saved to disk when daemon starts
 restore_persistence = False
+
+# Keyboard battery notifier
+keyboard_battery_notifier = True
+
+# Keyboard battery notification frequency [s] (0 to disable)
+keyboard_battery_notifier_freq = 600
```
Comment (z3ntu) [inline daemon/resources/razer.conf]: I don't think it makes much sense to keep keyboard and mouse separated. If it's not too much work please merge them, remove the `mouse_` prefix and keep a migration path so that `mouse_battery_notifier` and `mouse_battery_notifier_freq` don't stop working with this change.
Otherwise we can also keep it like this for now and merge the items them later.

**Review by z3ntu — CHANGES_REQUESTED** [review]
Thanks for the pull request! It looks quite good, I just have some mostly minor suggestions

File: daemon/resources/razer.conf (hunk 10 above)
Comment (IvanVojtko) [reply daemon/resources/razer.conf]: Done

File: install_files/udev/99-razer.rules (hunk 9 above)
Comment (IvanVojtko) [reply install_files/udev/99-razer.rules]: Done

File: driver/razerkbd_driver.h (hunk 8 above)
Comment (IvanVojtko) [reply driver/razerkbd_driver.h]: Done

File: driver/razerkbd_driver.h (hunk 7 above)
Comment (IvanVojtko) [reply driver/razerkbd_driver.h]: Done

File: driver/razerkbd_driver.c (hunk 6 above)
Comment (IvanVojtko) [reply driver/razerkbd_driver.c]: Done

File: driver/razerkbd_driver.c (hunk 5 above)
Comment (IvanVojtko) [reply driver/razerkbd_driver.c]: Done

File: driver/razerkbd_driver.c (hunk 3 above)
Comment (IvanVojtko) [reply driver/razerkbd_driver.c:342]: Yes it is interesting but that's what I got from packets and it works :D

File: daemon/openrazer_daemon/hardware/keyboards.py (hunk 2 above)
Comment (IvanVojtko) [reply daemon/openrazer_daemon/hardware/keyboards.py]: Done

File: README.md (hunk 1 above)
Comment (IvanVojtko) [reply README.md]: Done

**Comment by IvanVojtko** [comment]
@z3ntu all done!

**Review by z3ntu — APPROVED** [review]
Fixed up some remaining things, LGTM now

**Comment by nachitox** [comment]
Hi,
I thought it was something related to [polychromatic](https://github.com/polychromatic/polychromatic/issues/366) but it might be an openrazer issue.

I have this keyboard but
```
python3 -c "import openrazer.client; a = openrazer.client.DeviceManager(); print(a.devices[0].name, a.devices[0].has('battery'))"
```
returns
> Razer BlackWidow V3 Mini Hyperspeed (Wireless) False

Is that a bug?
